File: app/src/main/java/com/nimtego/plectrum/data/repository/datasource/DataStoreFactory.java
package com.nimtego.plectrum.data.repository.datasource;

import android.content.Context;

import com.nimtego.plectrum.data.cache.Cache;
import com.nimtego.plectrum.data.network.NetworkConnection;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.annotations.NonNull;

@Singleton
public class DataStoreFactory<E> {

    private final Context context;
    private final Cache<E> cache;

    @Inject
    public DataStoreFactory(@NonNull Context context, @NonNull Cache<E> cache) {
        this.context = context;
        this.cache = cache;
    }

    public DataStore create(int userId) {
        DataStore userDataStore;

        if (!this.cache.isExpired() && this.cache.isCached(String.valueOf(userId))) {
            userDataStore = new DiskDataStore(this.cache);
        } else {
            userDataStore = createCloudDataStore();
        }

        return userDataStore;
    }

    public DataStore createCloudDataStore() {

        return new CloudDataStore(NetworkConnection.getInstance(), this.cache);
    }
}

File: app/src/main/java/com/nimtego/plectrum/data/repository/datasource/DataStoreFactory.kt
package com.nimtego.plectrum.data.repository.datasource

import android.content.Context

import com.nimtego.plectrum.data.cache.Cache
import com.nimtego.plectrum.data.network.NetworkConnection

import javax.inject.Inject
import javax.inject.Singleton

import io.reactivex.annotations.NonNull

@Singleton
class DataStoreFactory<E>
@Inject
constructor(private val context: Context,
            private val cache: Cache<E>) {

    fun create(userId: Int): DataStore {
        val userDataStore: DataStore

        if (!this.cache.isExpired() && this.cache.isCached(userId.toString())) {
            userDataStore = DiskDataStore(this.cache)
        } else {
            userDataStore = createCloudDataStore()
        }

        return userDataStore
    }

    fun createCloudDataStore(): DataStore {

        return CloudDataStore(NetworkConnection.getInstance(), this.cache)
    }
}
